lib/Index/Index.cpp

```diff
@@ -440,6 +451,43 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
     return true;
   }
 
+  /// This method identifies containing declarations. Any references that are
+  /// associated with these declarations will obtain the \c RelationContainedBy
+  /// role.
+  ///
+  /// The use of \c PatternBindingDecl here ensures that variable declarations
+  /// will contain references to explicit types, initializers, custom
+  /// attributes, etc.
+  ///
+  /// Note that any single entry into the \c ContainerStack may have multiple
+  /// declarations, such is the case for a tuple, e.g:
+  /// let (a, b): (Int, String) = (x, y)
+  /// Currently both \c a and \c b will contain references to \c Int, \c String,
+  /// \c x and \c y. Future work could improve upon this to try match up the
+  /// indices, such that \c a would only contain references to \c Int and \c x,
+  /// and \c b references to \c String and \c y.
+  void beginBalancedDeclVisit(Decl *D) override {
+    if (auto *PBD = dyn_cast<PatternBindingDecl>(D)) {
+      SmallVector<Decl *, 6> VDs;
+      for (auto i : range(PBD->getNumPatternEntries())) {
+        PBD->getPattern(i)->forEachVariable(
+            [&](VarDecl *VD) { VDs.push_back(VD); });
+      }
+
+      if (VDs.size() > 0)
+        ContainerStack.push_back({VDs, D});
+    } else if (auto *AFD = dyn_cast<AbstractFunctionDecl>(D)) {
```

Comment:
I believe the previous check was needed because the `EntitiesStack` could contain reference entities. Whether you'd actually have a `AbstractFunctionDecl` on there that doesn't already have the `Definition` role, I'm not sure. For non-reference entities the `Definition` role is always applied: https://github.com/apple/swift/blob/main/lib/Index/Index.cpp#L1326

In the case of `ContainerStack` entities we don't actually call `initIndexSymbol` as it's not necessary, we don't create a `IndexSymbol` for them. These container declarations will exist both in `ContainerStack`, and in `EntitiesStack`. Those in the `EntitiesStack` _do_ have the associated `IndexSymbol`.

This was the reason I created the `ContainerStack`, as the `EntitiesStack` was already used for a multiple purposes: `getParentDecl` and `getContainingDecl` - which aren't necessarily the same thing. Trying to expand the scope of what could be considered a container declaration whilst still using `EntitiesStack` caused issues. It was just much simpler and cleaner to introduce a new stack solely for containers.